docs/auth.md

```diff
@@ -0,0 +1,112 @@
+
+* FerretDB should handle the client's authentication commands and use provided credentials to authenticate in PostgreSQL.
+
+* Implement createUser, dropUser, etc. commands.
+Update listCommands to include requiresAuth.
+Update connectionStatus to include authInfo.
+
+# Overview of the authentication commands
+
+## Useful links
+
+[MongoDB auth docs](https://github.com/mongodb/mongo/blob/a06bc8bbced8f0c60b94ed784f5f105f2f01ed5d/src/mongo/db/auth/README.md)
+
+[Authentication reference](https://www.mongodb.com/docs/manual/core/authentication/)
+[SCRAM authentication](https://www.mongodb.com/docs/manual/core/security-scram/)
+
+## Auth Commands
+
+| #   | command      | Description                        |
+|-----|--------------|------------------------------------|
+| 1   | authenticate | Authenticate with x.509 mechanism. |
+| 2   | saslStart    | Start SASL authentication.         |
+| 3   | saslContinue | Continue SASL authentication.      |
+
+
+## Users Management Commands
+
+| #   | command                  | Description                            |
+|-----|--------------------------|----------------------------------------|
+| 1   | createUser               | Creates user.                          |
+| 2   | dropAllUsersFromDatabase | Deletes all users from a database.     |
+| 3   | dropUser                 | Remove a single user from database.    |
+| 4   | grantRolesToUser         | Grants roles and privileges to a user. |
+| 5   | revokeRolesFromUser      | Removes roles from a user.             |
+| 6   | updateUser               | Updates a user's data.                 |
+| 7   | usersInfo                | Returns information about the users.   |
+
+## Commands that require authentication
+
+| #   | command          | Description                 |
+|-----|------------------|-----------------------------|
+| 1   | listCommands     | Returns a list of commands. |
+| 2   | connectionStatus | Returns connection status.  |
+
+
+# Tasks
+## Add support for separate connection pools for each user 
+
+*We will try to not store the username and password*.
+We should try to use PostreSQL's authentication mechanism.
+
+When a client connects with credentials specified in the connection string (username and password), authentication message would be sent to the server (`saslStart`).
+In order to create a new connection pool, we should use the username from the authentication message.
+
+When a client disconnects, we should close the connection pool.
+We should also close the system pool and all user pools when the server is shut down.
+
+To separate connection pools we should:
+* Use current connection pool as the `system` pool (we will use it later to query the database for user's credentials).
+* Handle `saslStart` message.
+* To distinguish between users we should add to the Handler instance a map of connection pools, where the key is the username.
+
+### Questions:
+* How to distinguish between users connected without authentication?
```

Comment:
Will it be possible still to connect without authentication? I thought it was a system-wide requirement setting to authenticate.

Reply:
No, it would not be possible to connect without authentication if it's enabled in PostgreSQL.

Reply:
> No, it would not be possible to connect without authentication if it's enabled in PostgreSQL.

So if `trusted` is enabled on Postgres, and auth enabled, then the user cannon login?

Reply:
`trusted` means PostgreSQL server trusts everyone, which means authentication is disabled. When password authentication is enabled username and a password are required to authenticate.